github-actions/utils/checkout-and-setup-node/action.yml

```diff
@@ -0,0 +1,48 @@
+name: 'Checkout and Setup Node'
+description: 'Checks out the repository and sets up node environment.'
+author: 'Angular'
+inputs:
+  fetch-depth:
+    default: 0
+    description: |
+      The number of commits to include in the fetch, defaulting to all commits in the history
+      of the repository
+
+  node-version:
+    description: |
+      A specific version of node to use for the node environment, exclusive with
+      node-version-file-path input.
+
+  node-version-file-path:
+    default: '.nvmrc'
+    description: |
+      Relative path to the nvm version file to set node version, exclusive with
+      node-version-file-path input.  Defaults to .nvmrc
+
+  cache-node-modules:
+    type: boolean
+    description: |
+      Whether to cache the node_modules directory, this is helpful when postinstalls are performed
+      and modify the resulting node_modules directory.  Defaults to False.
+
+runs:
+  using: composite
+  steps:
+    - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # tag=v3.5.3
+      with:
+        fetch-depth: ${{ inputs.fetch-depth }}
+        persist-credentials: false
+
+    - uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # tag=v3.6.0
+      with:
+        node-version-file: ${{ inputs.node-version-file-path }}
+        node-version: ${{ inputs.node-version }}
+        cache: 'yarn'
+
+    - uses: actions/cache@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # tag=v3.3.1
```

Comment:
So, this action manages the caching of node modules, but we don't install them as part of this action right? That might be a little too magical. WDYT?

Reply:
Correct we don't install them.  This is just about setting up the environment, not performing the actual installs and builds and such.

Reply:
I like the action overall, but just want to make sure we understand this. Do you think it's worth doing that in here? I'm imagining a complicated workflow where we'd want to cache other things too, or e.g. bazelisk, or baze repositories? Would we have multiple cache archives? (as we know from Circle- caching is not free).

I'm a little unhappy that we restore cache here, but the package manager needs to be invoked manually.

We can fine-tune afterwards, so LGTM 😄

Reply:
Lets fine tune as we go for sure, but I think that these multiple caches are worthwhile as they are not always managed by the same key.

For instance, caching bazel (bazelisk) and node_modules, are based on different things that don't change at the same time. So it would be better to have them in separate caches. The overhead/speed on Github actions seems much better than Circle